### PR TITLE
Feature/delete offering

### PR DESCRIPTION
**Pull request by fbielejec**

### CODE REVIEW

**Comment by fbielejec** [comment]
Preliminary PR

File: src/name_bazaar/server/db.cljs [hunk 1]
```diff
@@ -163,7 +163,7 @@
                                    :min-length :max-length :name-position :min-end-time-now? :version :node-owner?
                                    :top-level-names? :sub-level-names? :exclude-node :exclude-special-chars?
                                    :exclude-numbers? :limit :offset :order-by :select-fields :root-name :total-count?
-                                   :bidder :winning-bidder :exclude-winning-bidder :finalized? :sold?]
+                                   :bidder :winning-bidder :exclude-winning-bidder :finalized? :sold? :exclude-unregistered?] :as search-parameters
```
Comment (madvas) [inline src/name_bazaar/server/db.cljs]: We don't need to pass this as param. It will be turned on always

File: src/name_bazaar/server/db.cljs [hunk 2]
```diff
@@ -188,7 +190,8 @@
                                               [:= :end-time nil]])
               sold? (merge-where [:and
                                   [:<> :new-owner nil]
-                                  [:<> :new-owner emergency-state-new-owner]])
+                                  [:<> :new-owner emergency-state-new-owner]
+                                  [:<> :new-owner unregistered-new-owner]])
```
Comment (madvas) [inline src/name_bazaar/server/db.cljs]: This shouldn't be under `sold?` condition

File: src/name_bazaar/server/db.cljs [hunk 3]
```diff
@@ -176,7 +176,9 @@
                      :from [:offerings]
                      :offset offset
                      :limit limit}
-              original-owner (merge-where [:= :original-owner (string/lower-case original-owner)])
+              original-owner (merge-where [:= :original-owner (string/lower-case original-owner)])      
+              exclude-unregistered? (merge-where [:and
+                                                  [:<> :price unregistered-price-wei]])
```
Comment (madvas) [inline src/name_bazaar/server/db.cljs]: Should also check for version

File: src/name_bazaar/shared/utils.cljs [hunk 4]
```diff
@@ -1,10 +1,21 @@
...
   (:require
+    [cljs.core.match :refer-macros [match]]
     [clojure.string :as string]
     [district0x.shared.big-number :as bn]
     [district0x.shared.utils :as d0x-shared-utils :refer [zero-address? zero-address]]
     [cljs-web3.core :as web3]))
 
+(def emergency-state-new-owner "0x000000000000000000000000000000000000dead")
+(def unregistered-new-owner    "0x00000000000000000000000000000000deaddead")
+(def unregistered-price-wei                                        63466346)
```
Comment (madvas) [inline src/name_bazaar/shared/utils.cljs]: Can you please use traditional alignment, for consistency with the rest of app

File: src/name_bazaar/shared/utils.cljs [hunk 5]
```diff
@@ -1,10 +1,21 @@
...
+(def buy-now-offering-min-version 2)
+(def auction-offering-min-version 100001)
```
Comment (madvas) [inline src/name_bazaar/shared/utils.cljs]: `buy-now-offering-unregister-min-version`
`auction-offering-unregister-min-version`

File: src/name_bazaar/shared/utils.cljs [hunk 6]
```diff
@@ -47,9 +58,18 @@
...
-                     :offering/finalized-on])
+(defn supports-unregister? [offering-type offering-version]
+  (match [offering-type]
+         [:auction-offering] (>= offering-version auction-offering-min-version)
+         [:buy-now-offering] (>= offering-version buy-now-offering-min-version)))
+
+(defn unregistered? [offering]
+  (let [new-owner (:offering/new-owner offering)
+        price (-> offering :offering/price bn/->number)]
+    (match [(= unregistered-new-owner new-owner) (= unregistered-price-wei price)]
```
Comment (madvas) [inline src/name_bazaar/shared/utils.cljs]: When checking price, should also check for old contract version

File: src/name_bazaar/ui/components/offering/bottom_section.cljs [hunk 7]
```diff
@@ -33,9 +33,18 @@
        (dissoc props :offering))
      "Transfer Ownership"]))
 
+(defn delete-offering-button [{:keys [:address :offering-type] :as props}]
+  [transaction-button
+   {:color :violet
+    :pending-text "Deleting..."
+    :pending? @(subscribe [:offering.unregister/tx-pending? address offering-type])
+    :on-click #(dispatch [:offering/unregister {:offering/address address}])}
+   "Delete Offering"])
```
Comment (madvas) [inline src/name_bazaar/ui/components/offering/bottom_section.cljs]: only "Delete"

File: src/name_bazaar/ui/components/offering/bottom_section.cljs [hunk 8]
```diff
@@ -33,9 +33,18 @@
        (dissoc props :offering))
      "Transfer Ownership"]))
 
+(defn delete-offering-button [{:keys [:address :offering-type] :as props}]
+  [transaction-button
+   {:color :violet
```
Comment (madvas) [inline src/name_bazaar/ui/components/offering/bottom_section.cljs]: `:pink` ;)

File: src/name_bazaar/ui/components/offering/bottom_section.cljs [hunk 9]
```diff
@@ -63,7 +72,9 @@
            :disabled (not editable?)
            :color "purple"
            :href (path-for :route.offerings/edit {:offering/address address})}
-          "Edit"])])))
+          "Edit"])
+       (when can-delete?
+         [delete-offering-button {:address address :offering-type type}])])))
```
Comment (madvas) [inline src/name_bazaar/ui/components/offering/bottom_section.cljs]: rename keys to `:offering/address`, `:offering/type` for consistency with rest of app

File: src/name_bazaar/ui/components/offering/list_item.cljs [hunk 10]
```diff
@@ -132,7 +133,7 @@
      [offering-header-missing-ownership-tag
       {:offering offering}])
 
-   (when show-sold?
+   (when (and show-sold? (not (:offering/unregistered? offering)))
```
Comment (madvas) [inline src/name_bazaar/ui/components/offering/list_item.cljs]: Please put `(not (:offering/unregistered? offering))` condition inside `offering-header-sold-tag`

File: src/name_bazaar/ui/events/offerings_events.cljs [hunk 11]
```diff
@@ -240,6 +240,49 @@
...
+      {:dispatch [:district0x/make-transaction
+                  (merge
+                   {:name (gstring/format "Unregister %s offering" name)
+                    :tx-opts {:gas 200000 :gas-price default-gas-price}   
+                    :contract-key type
+                    :contract-address offering-address
+                    :form-id (select-keys form-data [:offering/address])
+                    :result-href (path-for :route.offerings/detail form-data)                   
+                    :on-tx-receipt-n [[:district0x.snackbar/show-message
+                                       (gstring/format "Unregistered offering %s" name)]
```
Comment (madvas) [inline src/name_bazaar/ui/events/offerings_events.cljs]: "Offering for %s was deleted"

File: src/name_bazaar/ui/events/offerings_events.cljs [hunk 12]
```diff
@@ -240,6 +240,49 @@
...
+  :offering/unregister
+  [interceptors (validate-first-arg (s/keys :req [:offering/address]))]
+  (fn [{:keys [:db]} [form-data]]
+    (let [offering-address (:offering/address form-data)
+          {:keys [:offering/type :offering/name :offering/version :offering/price
+                  :auction-offering/end-time :auction-offering/extension-duration :auction-offering/min-bid-increase] :as offering}
+          (get-offering db offering-address)]      
+      {:dispatch [:district0x/make-transaction
+                  (merge
+                   {:name (gstring/format "Unregister %s offering" name)
```
Comment (madvas) [inline src/name_bazaar/ui/events/offerings_events.cljs]: "Delete %s offering"

File: src/name_bazaar/ui/events/offerings_events.cljs [hunk 13]
```diff
@@ -495,7 +538,9 @@
        :dispatch [:offerings/search {:search-results-path search-results-path
                                      :append? (:append? opts)
                                      :on-success [:offerings/load {:load-ownership? (not (:finalized? search-params))}]
-                                     :params search-params}]})))
+                                     ;; TODO: or just exclude in db/search-offerings
+                                     :params (-> search-params
+                                                 (assoc :exclude-unregistered? true))}]})))
```
Comment (madvas) [inline src/name_bazaar/ui/events/offerings_events.cljs]: no need for this param, as stated above

File: src/name_bazaar/server/contracts_api/auction_offering.cljs [hunk 14]
```diff
@@ -67,6 +67,18 @@
                                          :from (state/active-address server-state)}
                                         opts)))
 
+(defn unregister! [server-state contract-address opts]
```
Comment (madvas) [inline src/name_bazaar/server/contracts_api/auction_offering.cljs]: can put this into `name-bazaar.server.contracts-api.offering`

File: src/name_bazaar/server/contracts_api/buy_now_offering.cljs [hunk 15]
```diff
@@ -32,6 +32,18 @@
                                          :from (state/active-address server-state)}
                                         opts)))
 
+(defn unregister! [server-state contract-address opts]
```
Comment (madvas) [inline src/name_bazaar/server/contracts_api/buy_now_offering.cljs]: can put this into `name-bazaar.server.contracts-api.offering`

File: resources/public/contracts/src/Offering.sol [hunk 16]
```diff
@@ -81,6 +81,14 @@ contract Offering {
     }
 
     /**
```
Comment (fbielejec) [inline resources/public/contracts/src/Offering.sol]: maybe we ought to add a meta-info map to the Offering contract for future updates?

**Pull request merged**